pkg/network/infraconfigurators/bridge_test.go

```diff
@@ -152,4 +152,521 @@ var _ = Describe("Bridge infrastructure configurator", func() {
+			mac                net.HardwareAddr
+			podLink            *netlink.GenericLink
+			podIP              netlink.Addr
+			vmi                *v1.VirtualMachineInstance
+		)
+
+		BeforeEach(func() {
+			iface = v1.DefaultBridgeNetworkInterface()
+			vmi = newVMIBridgeInterface("default", "vm1")
+
+			bridgeConfigurator = NewBridgePodNetworkConfigurator(vmi, iface, bridgeIfaceName, launcherPID, handler)
+		})
+
+		BeforeEach(func() {
+			mac, _ = net.ParseMAC(macStr)
+			podLink = &netlink.GenericLink{LinkAttrs: netlink.LinkAttrs{Name: ifaceName, MTU: mtu}}
+		})
+
+		When("the pod features an L3 configured network (IPAM)", func() {
+			var (
+				dummySwap              *netlink.Dummy
+				podLinkAfterNameChange *netlink.GenericLink
+			)
+
+			BeforeEach(func() {
+				bridgeIPStr = fmt.Sprintf(bridgeFakeIP, 0)
+				bridgeIPAddr, _ = netlink.ParseAddr(bridgeIPStr)
+				dummySwap = &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: ifaceName}}
+				inPodBridge = &netlink.Bridge{LinkAttrs: netlink.LinkAttrs{Name: bridgeIfaceName}}
+				podIP = netlink.Addr{IPNet: &net.IPNet{IP: net.IPv4(10, 35, 0, 6), Mask: net.CIDRMask(24, 32)}}
+				podLinkAfterNameChange = &netlink.GenericLink{LinkAttrs: netlink.LinkAttrs{Name: generateDummyIfaceName(ifaceName), MTU: mtu}}
+			})
+
+			BeforeEach(func() {
+				bridgeConfigurator.podNicLink = podLink
+				bridgeConfigurator.tapDeviceName = tapDeviceName
+				bridgeConfigurator.ipamEnabled = true
+				bridgeConfigurator.podIfaceIP = podIP
+			})
+
+			When("all network driver calls succeed", func() {
+				BeforeEach(func() {
+					handler.EXPECT().LinkSetDown(podLink).Return(nil)
+					handler.EXPECT().LinkAdd(inPodBridge).Return(nil)
+					handler.EXPECT().LinkSetMaster(podLinkAfterNameChange, inPodBridge).Return(nil)
+					handler.EXPECT().LinkSetUp(inPodBridge).Return(nil)
+					handler.EXPECT().ParseAddr(bridgeIPStr).Return(bridgeIPAddr, nil)
+					handler.EXPECT().AddrAdd(inPodBridge, bridgeIPAddr).Return(nil)
+					handler.EXPECT().DisableTXOffloadChecksum(inPodBridge.Name).Return(nil)
+					handler.EXPECT().CreateTapDevice(tapDeviceName, queueCount, launcherPID, mtu, netdriver.LibvirtUserAndGroupId).Return(nil)
+					handler.EXPECT().BindTapDeviceToBridge(tapDeviceName, inPodBridge.Name).Return(nil)
+					handler.EXPECT().LinkSetUp(podLinkAfterNameChange).Return(nil)
+					handler.EXPECT().LinkSetLearningOff(podLinkAfterNameChange).Return(nil)
+					handler.EXPECT().AddrDel(podLink, &podIP).Return(nil)
+					handler.EXPECT().LinkSetName(podLink, generateDummyIfaceName(ifaceName))
+					handler.EXPECT().LinkByName(generateDummyIfaceName(ifaceName)).Return(podLinkAfterNameChange, nil)
+					handler.EXPECT().LinkAdd(dummySwap).Return(nil)
+					handler.EXPECT().AddrReplace(dummySwap, &podIP).Return(nil)
+					handler.EXPECT().SetRandomMac(generateDummyIfaceName(ifaceName)).Return(mac, nil)
+					handler.EXPECT().ConfigureIpv4ArpIgnore().Return(nil)
+				})
+
+				It("should work", func() {
```

Comment:
If you skip reading the full context I agree it is not descriptive. 

As most things, when read in the proper context, they make sense: 

"Bridge infrastructure configurator,  prepare the pod networking infrastructure, when the pod features an L3 configured network (IPAM), when all network driver calls succeed, it should work". 

I'm open to suggestions; what would you name it then ?

Reply:
> The `It` is the actual test and all the rest are fixtures of the test. A test is suppose to express what it does clearly and not require the reader to travel hundreds of lines to understand it by composing its fixtures hierarchy names.

The user can instead just run the tests and read the output to get the full string.

> 
> If you reduce the hirarchy, it will read much better and will not require name tests that do not stand by themselves. Not all setups/fixtures must come as individual containers.
> 
> Here is one example: `"all network driver calls succeed"` can just be a helper function, called the same. I do not need to delve into how this is implemented when reading the test, I am mainly interested in knowing what was done. The details can be abstracted away.
> And it surely does not fit a proper test name (if you choose to see it as a composition of the containers). Saying something like "When IPAM is enabled, network preparation succeeds" is much clearer.
> (unlike with the success case, on failure you do need to explain the reason why it failed).

IMO, the string you just used as an example is a clear indication you should be using a `When` clause - encapsulating the steps of the test setup.

I like the suggestion of using a builder pattern to mock the handler; I'll use it, and see where it leads.

Reply:
Done.